### PR TITLE
Retry requests within retry count limit OR grace period (default 10s)

### PR DESCRIPTION
**Pull request by jonmv**
@bjorncs

### CODE REVIEW

File: vespa-feed-client/src/main/java/ai/vespa/feed/client/impl/HttpRequest.java [hunk 1]
```diff
@@ -37,8 +40,8 @@ public byte[] body() {
         return body;
     }
 
-    public Duration timeout() {
-        return timeout;
+    public Duration timeLeft() {
+        return Duration.ofMillis(deadlineMillis - clock.getAsLong());
```
Comment (bjorncs) [inline vespa-feed-client/src/main/java/ai/vespa/feed/client/impl/HttpRequest.java:44]: Consider using nano time to more accurately measure elapsed time (that is not affected by wall clock adjustments etc).

File: vespa-feed-client/src/main/java/ai/vespa/feed/client/impl/JettyCluster.java [hunk 2]
```diff
@@ -81,8 +82,11 @@ public void dispatch(HttpRequest req, CompletableFuture<HttpResponse> vessel) {
             Endpoint endpoint = findLeastBusyEndpoint(endpoints);
             try {
                 endpoint.inflight.incrementAndGet();
-                long reqTimeoutMillis = req.timeout() != null
-                        ? req.timeout().toMillis() * 11 / 10 + 1000 : IDLE_TIMEOUT.toMillis();
+                long reqTimeoutMillis = req.timeLeft().toMillis();
+                if (reqTimeoutMillis <= 0) {
+                    vessel.completeExceptionally(new TimeoutException("Operation timed out"));
```
Comment (bjorncs) [inline vespa-feed-client/src/main/java/ai/vespa/feed/client/impl/JettyCluster.java:87]: Consider adding timeout value to message (handy when user relies on default timeout value).

**Review by bjorncs — DISMISSED** [review]

File: vespa-feed-client/src/main/java/ai/vespa/feed/client/impl/GracePeriodCircuitBreaker.java [hunk 3]
```diff
@@ -80,21 +81,21 @@ public void failure(Throwable cause) {
...
     public State state() {
-        long failingMillis = clock.getAsLong() - failingSinceMillis.get();
-        if (failingMillis > graceMillis && halfOpen.compareAndSet(false, true))
+        long failingNanos = nanoClock.getAsLong() - failingSinceNanos.get();
+        if (failingNanos > graceNanos && halfOpen.compareAndSet(false, true))
             log.log(INFO, "Circuit breaker is now half-open, as no requests have succeeded for the " +
-                          "last " + failingMillis + "ms. The server will be pinged to see if it recovers" +
-                          (doomMillis >= 0 ? ", but this client will give up if no successes are observed within " + doomMillis + "ms" : "") +
+                          "last " + failingNanos / 1000 + "ms. The server will be pinged to see if it recovers" +
+                          (doomNanos >= 0 ? ", but this client will give up if no successes are observed within " + doomNanos / 1000 + "ms" : "") +
```
Comment (jonmv) [inline vespa-feed-client/src/main/java/ai/vespa/feed/client/impl/GracePeriodCircuitBreaker.java:94]:
```suggestion
                          "last " + failingNanos / 1_000_000 + "ms. The server will be pinged to see if it recovers" +
                          (doomNanos >= 0 ? ", but this client will give up if no successes are observed within " + doomNanos / 1_000_000 + "ms" : "") +
```

File: vespa-feed-client/src/main/java/ai/vespa/feed/client/impl/GracePeriodCircuitBreaker.java [hunk 4]
```diff
@@ -80,21 +81,21 @@ public void failure(Throwable cause) {
...
                           ". First failure was '" + detail.get() + "'.");
 
-        if (doomMillis >= 0 && failingMillis > doomMillis && open.compareAndSet(false, true))
-            log.log(WARNING, "Circuit breaker is now open, after " + doomMillis + "ms of failing request, " +
+        if (doomNanos >= 0 && failingNanos > doomNanos && open.compareAndSet(false, true))
+            log.log(WARNING, "Circuit breaker is now open, after " + doomNanos / 1000 + "ms of failing request, " +
```
Comment (jonmv) [inline vespa-feed-client/src/main/java/ai/vespa/feed/client/impl/GracePeriodCircuitBreaker.java:98]:
```suggestion
            log.log(WARNING, "Circuit breaker is now open, after " + doomNanos / 1_000_000 + "ms of failing request, " +
```

**Review by bjorncs — APPROVED** [review]

**Pull request merged**